### PR TITLE
#7128 JSON to Grok Converter

### LINKED ISSUE #7128: Conversion tool for ingest pipelines
## Overview

Elasticsearch [ingest node](https://www.elastic.co/guide/en/elasticsearch/reference/master/pipeline.html) feature is a subset of Logstash functionality to transform documents before it hits ES index. Ingest node pipeline is created using JSON structure. Logstash pipeline is created using a DSL config language.

## Migrating from ingest to Logstash

Users would want to migrate to Logstash for the following reasons:

* Take advantage of the richer transformation capabilities in Logstash. Ingest processors today is a subset (plus some additional ones exclusively in ingest) of what Logstash filters provide.
* Use multiple outputs. Ingest was designed to only support ES as an output. For example, users may want to archive their incoming data to S3 as well as indexing it in ES.
* Add complex data flow logic to their ingestion. For example, users may want to use conditionals in LS.
* Use the persistent queue feature to handle spikes when ingesting data (from beats and other sources).
* Use the JDBC/web service lookup plugins in Logstash to enrich data.

## Migration Tool

Our goal is to provide a 1-way migration path from ingest node json to Logstash for the reasons mentioned above. This can be achieved using a tool that transpiles ingest DSL into the LS config equivalent.

If the tool is written in javascript, we can use it in Kibana to provide a UI. Also, this tool can be run on Java 8 using [Nashhorn](http://www.oracle.com/technetwork/articles/java/jf14-nashorn-2126515.html). Hence with shared javascript code, we can provide a CLI tool using Java, and then a UI for easy migration.

### Pipeline-level rules:

1. Dots in field names: Ingest uses dots to access nested fields. For example: `apache2.access.remote_ip` needs to be converted to `[apache2][access][remote_ip]`.
2. `processors` map to `filters` in LS

### Filter level rules:

Each processor in ingest needs to be mapped to a filter in LS.

As a first step, we should target 1 ingest processor and convert them to LS. So create a grok pipeline in ingest and convert to LS pipeline with grok filter and no inputs and outputs.

For example:

Ingest JSON
```
{
   "description":"Pipeline for parsing apache error logs",
   "processors":[
      {
         "grok":{
            "field":"message",
            "patterns":[
               "%{IPORHOST:apache2.access.remote_ip} - %{DATA:apache2.access.user_name} \\[%{HTTPDATE:apache2.access.time}\\] \"%{WORD:apache2.access.method} %{DATA:apache2.access.url} HTTP/%{NUMBER:apache2.access.http_version}\" %{NUMBER:apache2.access.response_code} (?:%{NUMBER:apache2.access.body_sent.bytes}|-)( \"%{DATA:apache2.access.referrer}\")?( \"%{DATA:apache2.access.agent}\")?",
               "%{IPORHOST:apache2.access.remote_ip} - %{DATA:apache2.access.user_name} \\[%{HTTPDATE:apache2.access.time}\\] \"-\" %{NUMBER:apache2.access.response_code} -"
            ],
            "ignore_missing":true
         }
      }
   ]
}
```

Equivalent LS config:

```
filter {
   grok {
      match => { "message" => ["%{IPORHOST:[apache2][access][remote_ip]} - %{DATA:[apache2][access][user_name]} \[%{HTTPDATE:[apache2][access][time]}\] \"%{WORD:[apache2][access][method]} %{DATA:[apache2][access][url]} HTTP/%{NUMBER:[apache2][access][http_version]}\" %{NUMBER:[apache2][access][response_code]} (?:%{NUMBER:apache2.access.body_sent.bytes}|-)( \"%{DATA:[apache2][access][referrer]}\")?( \"%{DATA:[apache2][access][agent]}\")?",
        "%{IPORHOST:[apache2][access][remote_ip]} - %{DATA:[apache2][access][user_name]} \[%{HTTPDATE:[apache2][access][time]}\] \"-\" %{NUMBER:[apache2][access][response_code]} -" ] 
         }
     }
}


```

We can then generalize for the common filters in both ingest and LS.

### PR DESCRIPTION
**Pull request by original-brownbear**
Closes #7128 

Alright this is a version of the solution to #7128 I can stand by I think :)

* Separate module under `tools`
* Tried commenting the non-obvious js
* Tests and build run as part of the parent Gradle build

### CODE REVIEW

**Comment by original-brownbear** [comment]
@suyograo done :) I didn't expose any of the `js` logic yet. Let me comment on that in the main issue where you asked about it, give me a few minutes :)

**Comment by suyograo** [comment]
I think the top level gradle project should be called just `ingest-converter`. We can then have all the individual converters there. Because at some point the interface to the user will be 

```
bin/ingest-converter  --input ingest.json --output logstash.conf
```

It is a good idea to have `ingest` package as you have now, with individual converters. I would just shorten them to the name of processors: `tools/ingest-converter/src/main/java/org/logstash/ingest/Grok.java`

We can then add things like `Date.java` that will convert Date in ingest to Date in LS, `GeoIP` that will convert GeoIP processor to GeoIP in LS etc.

Let me know if this is confusing and we can jump on a zoom.

File: settings.gradle [hunk 1]
```diff
@@ -1,3 +1,4 @@
-include ':logstash-core', 'logstash-core-benchmarks'
+include ':logstash-core', 'logstash-core-benchmarks', 'ingest-json-to-grok'
```
Comment (suyograo) [inline settings.gradle]: Just make the top level as `ingest-converter` since we'll be dealing with more than just grok here.

File: tools/ingest-json-to-grok/build.gradle [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+javadoc {
+  enabled = true
+}
+
+apply plugin: 'com.github.johnrengelman.shadow'
+
+shadowJar {
+  baseName = 'ingest-json-to-logstash-grok'
```
Comment (suyograo) [inline tools/ingest-json-to-grok/build.gradle]: We can have just one jar `ingest-converter.jar` with the migration logic for all processors inside this.

File: tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import java.nio.file.Paths;
+import javax.script.Invocable;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+
+/**
+ * Ingest JSON DSL to Logstash Grok Transpiler.
+ */
+public final class JsonToGrok {
```
Comment (suyograo) [inline tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java]: Maybe just shorten this to Grok. We know we're in the ingest package at this point.

File: tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        // Utility Wrapper for JS Script.
+    }
+
+    public static void main(final String... args) throws ScriptException, NoSuchMethodException {
+        try (final Reader reader = new InputStreamReader(
+                     JsonToGrok.class.getResourceAsStream("/ingest-to-grok.js")
+             )
+        ) {
+            final ScriptEngine engine =
+                new ScriptEngineManager().getEngineByName("nashorn");
```
Comment (suyograo) [inline tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java]: Nice!

File: tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Ingest JSON DSL to Logstash Grok Transpiler.
+ */
+public final class JsonToGrok {
+
+    private JsonToGrok() {
+        // Utility Wrapper for JS Script.
+    }
+
+    public static void main(final String... args) throws ScriptException, NoSuchMethodException {
```
Comment (suyograo) [inline tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java]: I think it's great to have individual `main` entry points for each processor so we can test them individually. So +1 on keeping this even when we have a consolidated converter for all processors.

File: tools/ingest-json-to-grok/src/main/resources/ingest-to-grok.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                lines[i] = indent(lines[i], count);
+            } else {
+                lines[i] = indent(lines[i], count);
+            }
+        }
+        return lines.join("\n");
+    }
+
+    var parsed = JSON.parse(json);
+    var filters = parsed["processors"];
```
Comment (suyograo) [inline tools/ingest-json-to-grok/src/main/resources/ingest-to-grok.js]: call it `processors`

File: tools/ingest-json-to-grok/src/test/java/org/logstash/ingest/JsonToGrokTest.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public final class JsonToGrokTest {
```
Comment (suyograo) [inline tools/ingest-json-to-grok/src/test/java/org/logstash/ingest/JsonToGrokTest.java]: <3 the tests. Shorten to `GrokTest`

**Comment by original-brownbear** [comment]
@suyograo happy to hear that, just arrived at `PHL` will fix the points you mentioned in a sec :)

**Comment by original-brownbear** [comment]
@suyograo btw. if you have any examples of this conversion that aren't pretty much exactly the one I added, maybe we should add them as separate tests too? (test takes less than `1s`, so maybe it would be smart to get broader coverage while everything works as expected and before we extend/refactor?)

You could just paste the `json` for those here and I can go from there if you have any examples where you think it makes sense? :)

File: tools/ingest-json-to-grok/build.gradle (hunk 2 above)
Comment (original-brownbear) [inline tools/ingest-json-to-grok/build.gradle]: @suyograo makes sense :) renaming :)

File: settings.gradle (hunk 1 above)
Comment (original-brownbear) [inline settings.gradle]: @suyograo makes sense :) renaming :)

File: tools/ingest-json-to-grok/src/main/resources/ingest-to-grok.js (hunk 6 above)
Comment (original-brownbear) [inline tools/ingest-json-to-grok/src/main/resources/ingest-to-grok.js]: @suyograo makes sense :) renaming :)

File: tools/ingest-json-to-grok/src/test/java/org/logstash/ingest/JsonToGrokTest.java (hunk 7 above)
Comment (original-brownbear) [inline tools/ingest-json-to-grok/src/test/java/org/logstash/ingest/JsonToGrokTest.java]: @suyograo makes sense :) renaming :)

File: tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java (hunk 3 above)
Comment (original-brownbear) [inline tools/ingest-json-to-grok/src/main/java/org/logstash/ingest/JsonToGrok.java]: @suyograo makes sense :) renaming :)

**Comment by original-brownbear** [comment]
@suyograo renamings are in, let me know if you want to add further tests :)

**Comment by suyograo** [comment]
@original-brownbear good idea to add more tests. Here's a few. Please merge once you add these since the code LGTM

1. MySQL ingest
```
{
   "description":"Pipeline for parsing MySQL error logs,
   "processors":[
      {
         "grok":{
            "field":"message",
            "patterns":[
               "%{LOCALDATETIME:mysql.error.timestamp} (\\[%{DATA:mysql.error.level}\\] )?%{GREEDYDATA:mysql.error.message}",
               "%{DATA:mysql.error.timestamp} %{NUMBER:mysql.error.thread_id} \\[%{DATA:mysql.error.level}\\] %{GREEDYDATA:mysql.error.message1}",
               "%{GREEDYDATA:mysql.error.message2}"
            ],
            "ignore_missing":true,
            "pattern_definitions":{
               "LOCALDATETIME":"[0-9]+ %{TIME}",
               "GREEDYDATA1":".*"
            }
         }
      }
   ]
}
```

LS

```
filter {
   grok {
      match => { 
        "message" => ["%{LOCALDATETIME:[mysql][error][timestamp]} (\[%{DATA:[mysql][error][level]}\] )?%{GREEDYDATA:[mysql][error][message]}",
        "%{DATA:[mysql][error][timestamp]} %{NUMBER:[mysql][error][thread_id]} \[%{DATA:[mysql][error][level]}\] %{GREEDYDATA:[mysql][error][message1]}",
        "%{GREEDYDATA:[mysql][error][message2]}"] 
      }
      pattern_definitions => {
        "LOCALDATETIME" => "[0-9]+ %{TIME}"
        "GREEDYDATA1" => ".*"
      }
   }
}
```

2. Syslog ingest
```
{
   "description":"Syslog",
   "processors":[
      {
         "grok":{
            "field":"message",
            "patterns":[
               "%{SYSLOGTIMESTAMP:system.syslog.timestamp} %{SYSLOGHOST:system.syslog.hostname} %{DATA:system.syslog.program}(?:\\[%{POSINT:system.syslog.pid}\\])?: %{GREEDYMULTILINE:system.syslog.message}",
               "%{SYSLOGTIMESTAMP:system.syslog.timestamp} %{GREEDYMULTILINE:system.syslog.message}"
            ],
            "pattern_definitions":{
               "GREEDYMULTILINE":"(.|\n)*"
            },
            "ignore_missing":true
         }
      }
   ]
}
```

LS

```
filter {
   grok {
     match => {
      "message" => ["%{SYSLOGTIMESTAMP:[system][syslog][timestamp]} %{SYSLOGHOST:[system][syslog][hostname]} %{DATA:[system][syslog][program]}(?:\[%{POSINT:[system][syslog][pid]}\])?: %{GREEDYMULTILINE:[system][syslog][message]}",
              "%{SYSLOGTIMESTAMP:[system][syslog][timestamp]} %{GREEDYMULTILINE:[system][syslog][message]}"] 
     }
     pattern_definitions => { "GREEDYMULTILINE" => "(.|\n)*" }
   }
}

```

**Comment by suyograo** [comment]
We need to handle `pattern_definitions` field as well which is pretty similar in both implementations.

**Comment by original-brownbear** [comment]
@suyograo will you add an issue for the `pattern_definitions` situation or should I?

**Comment by original-brownbear** [comment]
@suyograo also since `pattern_definitions` are in all of your examples, maybe we should link here and add them once we have them correctly converting in the next issue? Otherwise this get's slightly confusing with having to adjust the test data in that step?

**Comment by suyograo** [comment]
Done https://github.com/elastic/logstash/issues/7241

**Comment by suyograo** [comment]
Good to merge, I'll take Date filter next. Do you want GeoIP? :)

**Comment by original-brownbear** [comment]
@suyograo sure just assigning me and I'll look into soon :)
